lib/pipeline.js

```diff
@@ -712,169 +701,148 @@ class PipelineModel extends BaseModel {
-                                    });
-                            }
-                            const job = updatedJobs.find(j => j.name === node.name);
+        const existingJobs = await this.jobs;
+
+        const jobsProcessed = [];
+        const updatedJobs = [];
+        const parsedConfigJobNames = Object.keys(parsedConfig.jobs);
+        const pipelineId = this.id;
+
+        // Loop through all existing jobs
+        for (const job of existingJobs) {
+            const jobName = job.name;
+            let requiresList = [];
+
+            // if it's in the yaml, update it
+            if (parsedConfigJobNames.includes(jobName)) {
+                const templateId = parsedConfig.jobs[jobName][0].templateId || null;
+
+                delete parsedConfig.jobs[jobName][0].templateId;
+
+                const permutations = parsedConfig.jobs[jobName];
 
-                            // Handle internal nodes
-                            if (job) {
-                                node.id = job.id;
+                requiresList = permutations[0].requires || [];
+                job.permutations = permutations;
+                job.templateId = templateId;
+                job.archived = false;
+                updatedJobs.push(await job.update());
+            } else if (!job.isPR()) {
+                job.archived = true;
+                updatedJobs.push(await job.update());
+            }
+
+            // sync external triggers for existing jobs
+            await syncExternalTriggers({
+                pipelineId,
+                jobName,
+                requiresList
+            });
+
+            // if it's a PR, leave it alone
+            jobsProcessed.push(job.name);
+        }
+
+        // Loop through all defined jobs in the yaml
+        for (const jobName of Object.keys(parsedConfig.jobs)) {
+            const permutations = parsedConfig.jobs[jobName];
+            const jobConfig = {
+                pipelineId,
+                name: jobName,
+                permutations
+            };
+            const requiresList = permutations[0].requires || [];
+
+            // If the job has not been processed, create it (new jobs)
+            if (!jobsProcessed.includes(jobName)) {
+                updatedJobs.push(await factory.create(jobConfig));
+
+                await syncExternalTriggers({
+                    pipelineId: this.id,
+                    jobName,
+                    requiresList
+                });
+            }
+        }
+
+        const { nodes } = this.workflowGraph;
+
+        // Add jobId to workflowGraph.nodes
+        await Promise.all(
+            nodes.map(node => {
+                // Handle external nodes
+                if (/sd@/.test(node.name)) {
+                    const pipelineFactory = this._getPipelineFactory();
+                    const [, externalPipelineId, externalJobName] = /sd@(\d+):([\w-]+)$/.exec(node.name);
```

Comment:
can this hardcoded regex be replaced with this one?
https://github.com/screwdriver-cd/data-schema/blob/master/config/regex.js#L60

Reply:
done.

Reply:
ah I meant line 83

Reply:
oops, fixed.